hummingbot/core/utils/api_throttler.py

```diff
@@ -0,0 +1,170 @@
+import asyncio
+import time
+
+from collections import deque
+from dataclasses import dataclass
+from enum import Enum
+from typing import (
+    Deque,
+    Dict,
+    List,
+)
+
+
+class RateLimitType(Enum):
```

Comment:
The different rate limits alternatives should be modeled as subclasses in a class hierarchy, polymorphic with each other. And in every place we should interact with them the same way.

Reply:
I think I understand what you mean here. We will have more classes to manage then but we get to transform the `weighted_task()`, `per_method_task()` into an abstract `execute_task()` function 🤔

Reply:
Yes, exactly. And that means that any connector requiring a particular throttler gets configured once with the correct instance, but the internal logic in any connector ends up interacting with the throttler in the same generic and polymorphic way.